### PR TITLE
(WIP) User location search radius should be geographically not as the crow flies

### PR DESCRIPTION
**Pull request by pezholio**
## Trello card URL:

https://trello.com/c/Zi8Eu8w9/513-user-location-search-radius-should-be-geographically-not-as-the-crow-flies

## Changes in this PR:

This PR uses the [Google Distance Matrix API](https://developers.google.com/maps/documentation/distance-matrix/start) to check if vacancies are really in an area before presenting them to a user. This can occur if a user is searching a large radius in somewhere like Cardiff where there are vacancies over the Bristol Channel in somewhere like Weston-Super-Mare. 

It works like so:

* User enters a location and radius
* We geocode the location, and search for schools within that radius as the crow flies (using the [GeoKit gem](https://github.com/geokit/geokit), which, under the hood, generates a Postgres query using the [Haversine formula](https://en.wikipedia.org/wiki/Haversine_formula)
* We then pass the user's location and the latlngs to the Distance Matrix API, filtering any schools that are not actually reachable by land
* We then pass the IDs of those schools to the ElasticSearch query, so we know that only vacancies at those schools will be returned

The only drawback with this is that we are now making an extra API call to Google when a user carries out a distance search, which may come with extra costs / a performance hit. Might be worth getting this up on Edge and see if we notice any change in Skylight.

### CODE REVIEW

File: app/models/school.rb [hunk 1]
```diff
@@ -31,19 +34,28 @@ def northing=(northing)
...
-      wgs84 = Breasal::EastingNorthing.new(
-        easting: easting.to_i,
-        northing: northing.to_i,
-        type: :gb
-      ).to_wgs84
+    self.latitude = wgs84[:latitude]
+    self.longitude = wgs84[:longitude]
+  end
+
+  def wgs84
```
Comment (erbridge) [inline app/models/school.rb]: What is `wgs84`?

File: app/views/hiring_staff/vacancies/_show.html.haml [hunk 2]
```diff
@@ -78,11 +78,11 @@
           %dd
             = link_to("#{@vacancy.school.name} website (opens in a new window)", @vacancy.school.url, class: 'govuk-link wordwrap', target: '_blank')
 
-      - if @vacancy.school.geolocation
+      - if @vacancy.school.latitude && @vacancy.school.longitude
```
Comment (erbridge) [inline app/views/hiring_staff/vacancies/_show.html.haml]: We check these a lot. Perhaps refactor to have `.geolocation?` or something?

File: db/migrate/20190408143457_add_lat_and_lng_to_schools.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class AddLatAndLngToSchools < ActiveRecord::Migration[5.2]
+  def change
+    add_column :schools, :latitude, :float
+    add_column :schools, :longitude, :float
+    remove_column :schools, :geolocation
```
Comment (erbridge) [inline db/migrate/20190408143457_add_lat_and_lng_to_schools.rb]: Do we need a task to migrate the data over, too?

Comment (pezholio) [reply db/migrate/20190408143457_add_lat_and_lng_to_schools.rb]: I think (@tahb, correct me if I'm wrong) we should be able to run the `data:schools:import` job on deploy, which will update the schools from the source data. The only issue is there may be a small window where location searches may not work properly.

**Comment by pezholio** [comment]
Gaah, so this hasn't made it very far. Just tried with a full corpus of schools, and we're already hitting limitations. There's a limit of 25 destinations per request, so even with a very small area, there's going to be more than 25 schools nearby, so I think this might not be a goer in its current form :(

**Comment by erbridge** [comment]
We could do requests in batches of 25, right? Any way to do the requests concurrently so we wouldn't have to wait for a round trip to trigger the next one?

**Comment by pezholio** [comment]
Yeah, we could do it in batches, but I'm worried that would slow things down quite considerably. I'm looking into other alternatives, like [Graphhopper](https://www.graphhopper.com) at the moment

**Comment by pezholio** [comment]
Right, so I've made some tweaks, so now we only fetch schools that actually have vacancies, rather than fetching all schools, which was a bit of a blunt instrument. I also had some problems with VCR, so have refactored the tests a bit too and removed the dependency on VCR that I introduced.

I'd still be keen to test this against a corpus of real live data on Edge or something. Not sure of the best way to do this though, as the real litmus test would be the amount of data.

File: Gemfile.lock [hunk 4]
```diff
@@ -468,6 +478,7 @@ GEM
     validate_url (1.0.6)
       activemodel (>= 3.0.0)
       public_suffix
+    vcr (4.0.0)
```
Comment (erbridge) [inline Gemfile.lock]: `vcr` is still here.

File: app/services/schools_in_area.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+class SchoolsInArea
+  attr_accessor :lat, :lng, :radius
```
Comment (erbridge) [inline app/services/schools_in_area.rb]: Should this be `attr_reader` since we're memoizing based on these values, so if they change, our data will become invalid?

File: app/services/vacancy_search_builder.rb [hunk 6]
```diff
@@ -79,15 +74,23 @@ def keyword_query
...
-        }
-      }
+      bool: {
+        filter: {
+          terms: {
+            school_id: schools_in_area,
+          },
+        },
+      },
     }
```
Comment (erbridge) [inline app/services/vacancy_search_builder.rb]: This looks a little weird to me. This results in something that looks like:
```
{
  bool: {
    must: [
      bool: {
        filter: {
          terms: {
            school_id: schools_in_area,
          }
        }
      },
      ...
    ]
  }
}
```
Whereas, `filter` used to be set on the first `bool`. Was that intentional?

File: app/views/vacancies/_vacancies_for_map.json.jbuilder [hunk 7]
```diff
@@ -1,7 +1,7 @@
-json.array! vacancies.select(&:school_geolocation) do |vacancy|
+json.array! vacancies.select do |vacancy|
```
Comment (erbridge) [inline app/views/vacancies/_vacancies_for_map.json.jbuilder]: Should this be filtering by ones that have `schoool.geolocation? == true`?

File: app/services/schools_in_area.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def matrix
+    @matrix ||= begin
+      matrix = GoogleDistanceMatrix::Matrix.new
+      matrix.origins << GoogleDistanceMatrix::Place.new(lat: lat, lng: lng)
+      schools.each do |s|
+        matrix.destinations << s.place
+      end
+      matrix
```
Comment (erbridge) [inline app/services/schools_in_area.rb]: Do we need to guard against there being more than 25? Perhaps log something so we hear about it, and fall back to a basic radius search? Or make more API calls?

File: spec/lib/schools_in_area_spec.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
+require 'rails_helper'
+
+RSpec.describe 'SchoolsInArea' do
```
Comment (erbridge) [inline spec/lib/schools_in_area_spec.rb]: We should probably also test what happens when there are more than 25 schools to search.

File: spec/presenters/school_vacancies_presenter_spec.rb [hunk 10]
```diff
@@ -35,7 +35,8 @@
 
     it 'returns sorted draft vacancies' do
       sort.update(column: 'job_title', order: 'desc')
-      expect(presenter.vacancies).to eq(draft_vacancies.sort_by { |vacancy| vacancy['job_title'] }.reverse)
+      vacancy_ids = draft_vacancies.sort_by { |vacancy| vacancy['job_title'] }.reverse.pluck(:id)
+      expect(presenter.vacancies.pluck(:id)).to eq(vacancy_ids)
```
Comment (erbridge) [inline spec/presenters/school_vacancies_presenter_spec.rb]: Think we could get these pulled out into their own PR?

File: app/services/schools_in_area.rb (hunk 5 above)
Comment (pezholio) [reply app/services/schools_in_area.rb]: Good catch 👍

File: spec/presenters/school_vacancies_presenter_spec.rb (hunk 10 above)
Comment (pezholio) [reply spec/presenters/school_vacancies_presenter_spec.rb]: Yep, we don't need this now 👍

File: app/services/schools_in_area.rb (hunk 8 above)
Comment (pezholio) [reply app/services/schools_in_area.rb]: Yep, I've done this now - breaking them down into batches of 25 and making subsequent calls. I'm still curious / concerned about how this will work in practice. Are we going to end up with really poor performance on areas with lots of schools?

File: app/services/vacancy_search_builder.rb (hunk 6 above)
Comment (pezholio) [reply app/services/vacancy_search_builder.rb]: I think (and correct me if I'm wrong) that we were searching by all the search criteria, and then filtering by distance after doing the initial search. Now we're doing the reverse and the school ids are now things to be searched on, rather than filtered by. My ElasticSearch knowledge is not the best, so I could be wrong...

File: app/views/vacancies/_vacancies_for_map.json.jbuilder (hunk 7 above)
Comment (pezholio) [reply app/views/vacancies/_vacancies_for_map.json.jbuilder]: Good catch. I dunno what made me delete that line. That said, I don't even thing we use this partial, so I've binned it

File: app/services/vacancy_search_builder.rb (hunk 6 above)
Comment (erbridge) [reply app/services/vacancy_search_builder.rb]: The difference between `must` and `filter` is only whether the score is important. `must` uses the score, while `filter` discards it. If they're on the same level, I believe they're run concurrently, and it's a bit of a simpler read than having nested `bool` queries, IMO. I _think_ the functionality is the same, though, but I'm not sure.

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html

File: app/views/vacancies/_vacancies_for_map.json.jbuilder (hunk 7 above)
Comment (erbridge) [reply app/views/vacancies/_vacancies_for_map.json.jbuilder]: :+1:

File: app/services/schools_in_area.rb (hunk 8 above)
Comment (erbridge) [reply app/services/schools_in_area.rb]: If we log something with the number of schools, or some note about having to batch them, maybe that would help us track if it's occuring at least?

**Comment by pezholio** [comment]
[User location search radius should be geographically not as the crow flies](https://trello.com/c/Zi8Eu8w9/513-user-location-search-radius-should-be-geographically-not-as-the-crow-flies)

**Comment by adrianclay** [comment]
Closing as this isn't a current priority.  Can re-open if we want to restart work on this.
